Fix for Invalidation of DeviceMapping Cache when Detaching Volumes

BugRoger opened the pull request:
There is a `deviceMapping` cache that keeps the state of the block device mappings in local memory. When a device is detached the cache does not get invalidated. A subsequent attachment for the same volume then falsely assumes that the device is already attached. It skip the actual API call to attach the volume and gets stuck in an endless loop waiting for the attachment to finish. It eventually times out and immediately starts to wait again. This only resolves once the kubelet is restarted.

This fix releases the device from the `deviceMapping` cache when a volume is detached. 

With this fix volume attach/detach operations work perfectly for me. I'm a bit unsure though how such a fundamental bug wasn't noticed before? Did this ever work and this is a regression? It could be that previous tests had the pods scheduled to a different kubelet with cold cache? Maybe @justinsb has some insight.

BugRoger committed:
Fix for Invalidation of DeviceMapping-Cache

justinsb commented:
Ouch - I think you are 100% right @BugRoger.  It would be great to have an integration test that detects this, but I think that getting this into 1.1 is more important.  (And likely cherry-picking to 1.0 also).

I think there's no real reason for it to be missing other than my mistake.  Sorry.

The handling of the case where there is a timeout is tricky.  It might be that the volume does eventually detach, or it might be that it never detaches.  I think it would also be cleaner if we could reuse releaseMountDevice.  But I think getting this fix in is more important.  If you have time I would like to see:

* a TODO comment to consider fully the timeout case
* a TODO comment to reuse releaseMountDevice
* a log-message if we call releaseVolume with a volume that isn't in the map (similar to releaseMountDevice)

If you don't have time, then no problem, I think this is bad enough that we should merge as is (the code looks great other than the above nits), and I can add those 3 nits later.

Thank you so much for tracking this down & providing a fix @BugRoger 

(@smarterclayton - fine to assign this one to me if you'd prefer).

BugRoger commented:
Alright, I addressed your comments.

In my case (private EC2 compatible cloud) the timeout of 60s is almost always too short and the detach operation times out. Fortunately, it does recover. When the volume gets attached again while it's still in `detaching` theres in an error/retry loop until it eventually succeeds (not sure it ever gives up). Though not nice, it does work for now.

justinsb commented:
Going to close & reopen to try to get shippable to re-run tests.

justinsb commented:
ok to test

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1192,6 +1192,24 @@ func (aws *AWSCloud) DetachDisk(instanceName string, diskName string) error {
+	// At this point we are waiting for the volume being detached. This
+	// releases the volume and invalidates the cache even when there is a timeout.
+	//
+	// TODO: A timeout leaves the cache in an inconsitent state. The volume is still
```
Typo nit: `s/inconsitent/inconsistent`

justinsb commented:
I don't know what is wrong with shippable.  I suggested a typo nit, which if you fix will have the nice side effect of getting shippable to try again :-)

BugRoger commented:
Unfortunately I found yet another invalidation bug. The cache is a map from single char to `volumeID`. As implemented here:

```
if strings.HasPrefix(mountpoint, "/dev/sd") {
    mountpoint = mountpoint[7:]
}
if strings.HasPrefix(mountpoint, "/dev/xvd") {
	mountpoint = mountpoint[8:]
}
deviceMappings[mountpoint] = orEmpty(blockDevice.EBS.VolumeID)
```

During the invalidation the whole mountpath is passed in:
```
ec2Device := "/dev/xvd" + mountpoint
defer func() {
	if !attached {
		awsInstance.releaseMountDevice(disk.awsID, ec2Device)
	}
}()
```

This never works and in case `AttachDisk` runs into a timeout, the kubelet yet in another endless loop. The fix is easy but there is still followup trouble here:
https://github.com/BugRoger/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1153

I even implemented that TODO. That also didn't fix the problem completely. To be quite honest I don't know where it still goes wrong, so I invalidate the whole cache upon timeouts in my fork:
https://github.com/BugRoger/kubernetes/commit/b9587354956c3ae4195eadf84f7fad1d1f401982 

To fix this properly I feel like we need (better) tests. Then refactor and model the state carefully, maybe with periodical reconciliation. That can go into a seperate PR, I hope.

justinsb commented:
This LGTM; going to close & reopen to kick shippable.

a-robinson commented:
Shippable has been incredibly slow today - I'm going to take @justinsb's last comment as an LGTM and merge. I assume that this should also be cherry-picked into 1.1?

a-robinson commented:
Before merging, would you mind squashing your commits, @BugRoger?

BugRoger commented:
@a-robinson squashed!

brendandburns commented:
@k8s-bot test this please

justinsb commented:
I think this is going to conflict with a parallel PR I did where I introduced a strongly-typed mountpoint, but let's merge this and I'll rebase.  I think it just needs an LGTM.  Going to close & reopen to get shippabel to pass first.

justinsb commented:
Poking shippable again

jsafrane commented:
Just for the reference, this may fix bug #15073

iterion commented:
Just saw this happen in a cluster I'm testing. Another workaround is to snapshot your volume and create a new one. Happy to see it will merge soon!

wojtek-t commented:
@k8s-bot unit test this please

wojtek-t commented:
@k8s-bot unit test this please

wojtek-t commented:
@k8s-bot unit test this please

justinsb commented:
@wojtek-t did I do something wrong?  Is there a new guide to what I have to do to get a PR assigned to me for review to actually merge once it looks good?

wojtek-t commented:
@justinsb - we have huge problems with Shippable now (it takes hours to finish often (I think ~50%) returns timeout)
So we just enable running our unit/integration test on Jenkins (you still have to manually trigger it as in: https://github.com/kubernetes/kubernetes/pull/14493#issuecomment-152176341).
Merge-bot need one of "Shippable" and "Unit test Jenkins" to be green (additionally to Travis and e2e Jenkins) to pick the PR up for merging.

FestivalBobcats commented:
Assuming this did not make the 1.1 release? It may not be related, but I'm seeing this after creating/deleting a few pods on a fresh 1.1 cluster:

```
"Error syncing pod, skipping: Error attaching EBS volume:
InvalidParameterValue: Value (/dev/xvd) for parameter device is invalid.
/dev/xvd is not a valid EBS device name."
```

roberthbailey commented:
@FestivalBobcats - it did not make the 1.1 release.

roberthbailey commented:
I just created a cherry pick of this PR into the 1.1 release branch (#18559). Once that gets merged, this should be resolved with the next patch release (1.1.4).

iameli commented:
If I get a pod/node into this state, is there an easy workaround I can use until 1.1.4 drops?

BugRoger commented:
> If I get a pod/node into this state, is there an easy workaround I can use until 1.1.4 drops?
Restart the Kubelet :frowning:

actionshrimp commented:
Just upgraded to 1.1.4 and still getting this issue - had a look in the release and it looks like the PR didn't make it in?

antoineco commented:
It should have. Did you build from the 1.1 release branch or are you using the beta?

actionshrimp commented:
Thanks for the quick reply. I looked at the source in the 1.1.4 release bundle here https://github.com/kubernetes/kubernetes/releases/tag/v1.1.4, and I'm using the binaries hosted on here: https://storage.googleapis.com/kubernetes-release/release/v1.1.4/bin/

FestivalBobcats commented:
I am also still seeing this issue on 1.1.4. It seems I can't ever re-attach the same volume to a node.

My team had been using 1.2.0-alpha3, and we had a frequent "/dev/xvdf is already in use" error. We thought this thread was relevant, but after building and deploying 1.1.4, we're now thinking that was a different error entirely.

On 1.1.4, we can't restart a pod with attached EBS without getting:

`Error syncing pod, skipping: Timeout waiting for volume state`

We're currently snapshotting and re-creating volumes as a workaround.

FestivalBobcats commented:
Correction: I'm completely wrong. Didn't build the right tag.

antoineco commented:
This did not fix #15073 unfortunately, I'm still in the same situation as before: detached volumes not attaching.

```
aws.go:1018] Waiting for volume state: actual=detached, desired=attached
```

roberthbailey commented:
@antoineco This wasn't cherry picked in time to catch the 1.1.4 release. You should see it fixed in the (pending) 1.1.7 release.

stepanstipl commented:
I still seem to have this (or very similar) issue in 1.1.7. When pod with EBS backed block storage is recreated, it will never succeed to attach the EBS volume:

```
FirstSeen     LastSeen        Count   From                                                    SubobjectPath   Reason          Message
  ─────────     ────────        ─────   ────                                                    ─────────────   ──────          ───────
  22m           22m             1       {scheduler }                                                            Scheduled       Successfully assigned monitoring-influxdb-v2-1-0q6zy to ip-172-20-217-186.eu-west-1.compute.internal
  20m           20m             1       {kubelet ip-172-20-217-186.eu-west-1.compute.internal}                  FailedMount     Unable to mount volumes for pod "monitoring-influxdb-v2-1-0q6zy_kube-system": Timeout waiting for volume state
  20m           20m             1       {kubelet ip-172-20-217-186.eu-west-1.compute.internal}                  FailedSync      Error syncing pod, skipping: Timeout waiting for volume state
  22m           5s              131     {kubelet ip-172-20-217-186.eu-west-1.compute.internal}                  FailedMount     Unable to mount volumes for pod "monitoring-influxdb-v2-1-0q6zy_kube-system": Error attaching EBS volume: VolumeInUse: vol-9b981f59 is already attached to an instance status code: 400, request id:
  22m           5s              131     {kubelet ip-172-20-217-186.eu-west-1.compute.internal}          FailedSync      Error syncing pod, skipping: Error attaching EBS volume: VolumeInUse: vol-9b981f59 is already attached to an instance status code: 400, request id:
```

Morriz commented:
I have tested this to still be an issue all the way up to 1.2.0-alpha.8. Can we get some more eyeballs on this?

Morriz commented:
I should be more specific: I don't get more than the timeout message:

```
FirstSeen	LastSeen	Count	From							SubobjectPath	Reason		Message
  ─────────	────────	─────	────							─────────────	──────		───────
  11m		11m		1	{scheduler }								Scheduled	Successfully assigned grafana-3n68f to ip-10-0-0-168.eu-central-1.compute.internal
  10m		28s		11	{kubelet ip-10-0-0-168.eu-central-1.compute.internal}			FailedMount	Unable to mount volumes for pod "grafana-3n68f_kube-system": Timeout waiting for volume state
  10m		28s		11	{kubelet ip-10-0-0-168.eu-central-1.compute.internal}			FailedSync	Error syncing pod, skipping: Timeout waiting for volume state
```

justinsb commented:
@Morriz can you post your pod description please?  I don't think the out-of-the-box grafana pod uses an EBS volume...

`kubectl describe pod --namespace=kube-system grafana-3n68f_kube-system -ojson`

Morriz commented:
sry, i am in a constant state of flux with the stack...trying to find the magic fit between components and their versions...so no backtracking there for now...maybe later

stepanstipl commented:
@justinsb What is interesting that even if you delete and recreate rc, and delete and recreate actual EBS volume (so you end up both with different pod and different EBS volume id) it will still keep failing. I should be able to post here description of actual failing pod tomorrow or on Monday.

In case this would be useful in the meantime the only difference compared to out of the box grafana is this in the rc definition, where the one I'm using adds:
```
        volumeMounts:
        - name: secret-volume
          readOnly: true
          mountPath: /secrets
        - name: grafana-persistent-storage
          mountPath: /var/lib/grafana
      volumes:
      - name: secret-volume
        secret:
          secretName: monitoring-grafana
      - name: grafana-persistent-storage
        awsElasticBlockStore:
          volumeID: "aws://$AWS_ZONE/$AWS_EBS_ID"
          fsType: ext4
```

But although coincidentally in this case it's also happening with grafana pod, I have had this issue on other pods with different images using EBS volumes - for example with influxdb or elasticsearch images.

Morriz commented:
@stepanstipl: that grafana ebs volume id has the old format, it is now just the volume-id (vol-1234567). Hope that was not causing the mount failure :)

Morriz commented:
@justinsb: I had only changed the default emptyDir storage to an ebs volume. Now trying to see under what conditions the mounts fail...

jtbcgdv commented:
@justinsb I have the same problem on v1.1.8.
Let me know if you need any more info

```
    Image ID:		
    QoS Tier:
      cpu:	Burstable
    Limits:
      cpu:	250m
    Requests:
      cpu:		150m
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Environment Variables:
Conditions:
  Type		Status
  Ready 	False 
Volumes:
  jenkins:
    Type:	AWSElasticBlockStore (a Persistent Disk resource in AWS)
    VolumeID:	aws://eu-west-1b/vol-3ca8ddf2
    FSType:	ext4
    Partition:	0
    ReadOnly:	false
  ssl:
    Type:	Secret (a secret that should populate this volume)
    SecretName:	jenkins-ssl
  default-token-bd5uj:
    Type:	Secret (a secret that should populate this volume)
    SecretName:	default-token-bd5uj
Events:
  FirstSeen	LastSeen	Count	From							SubobjectPath	Reason		Message
  ─────────	────────	─────	────							─────────────	──────		───────
  11m		11m		1	{scheduler }								Scheduled	Successfully assigned jenkins-leader-i9kk1 to ip-10-0-0-94.eu-west-1.compute.internal
  9m		27s		10	{kubelet ip-10-0-0-94.eu-west-1.compute.internal}			FailedMount	Unable to mount volumes for pod "jenkins-leader-i9kk1_default": Timeout waiting for volume state
  9m		27s		10	{kubelet ip-10-0-0-94.eu-west-1.compute.internal}			FailedSync	Error syncing pod, skipping: Timeout waiting for volume state
```

Morriz commented:
James, you are using the old volume id notation, which was superseded by just the aws notation ('vol-1234abcd'), and which is reflected in the kubernetes documentation as well....

> On 29 feb. 2016, at 12:54, James Thimont <notifications@github.com> wrote:
> 
> @justinsb <https://github.com/justinsb> I have the same problem on v1.1.8.
> Let me know if you need any more info
> 
>     Image ID:       
>     QoS Tier:
>       cpu:  Burstable
>     Limits:
>       cpu:  250m
>     Requests:
>       cpu:      150m
>     State:      Waiting
>       Reason:       ContainerCreating
>     Ready:      False
>     Restart Count:  0
>     Environment Variables:
> Conditions:
>   Type      Status
>   Ready     False 
> Volumes:
>   jenkins:
>     Type:   AWSElasticBlockStore (a Persistent Disk resource in AWS)
>     VolumeID:   aws://eu-west-1b/vol-3ca8ddf2
>     FSType: ext4
>     Partition:  0
>     ReadOnly:   false
>   ssl:
>     Type:   Secret (a secret that should populate this volume)
>     SecretName: jenkins-ssl
>   default-token-bd5uj:
>     Type:   Secret (a secret that should populate this volume)
>     SecretName: default-token-bd5uj
> Events:
>   FirstSeen LastSeen    Count   From                            SubobjectPath   Reason      Message
>   ─────────   ────────    ───── ────                            ───────────── ──────      ───────
>   11m       11m     1   {scheduler }                                Scheduled   Successfully assigned jenkins-leader-i9kk1 to ip-10-0-0-94.eu-west-1.compute.internal
>   9m        27s     10  {kubelet ip-10-0-0-94.eu-west-1.compute.internal}           FailedMount Unable to mount volumes for pod "jenkins-leader-i9kk1_default": Timeout waiting for volume state
>   9m        27s     10  {kubelet ip-10-0-0-94.eu-west-1.compute.internal}           FailedSync  Error syncing pod, skipping: Timeout waiting for volume state
> —
> Reply to this email directly or view it on GitHub <https://github.com/kubernetes/kubernetes/pull/14493#issuecomment-190179033>.
>
